src/music_gui.cpp

```diff
@@ -35,278 +36,388 @@
+void MusicSystem::ChangeMusicSet(const std::string &set_name)
 {
-	while (*playlist != 0 && playlist <= last) {
-		/* Song indices are saved off-by-one so 0 is "nothing". */
-		if (*playlist <= NUM_SONGS_AVAILABLE && !StrEmpty(GetSongName(*playlist - 1))) {
-			playlist++;
-			continue;
-		}
-		for (byte *p = playlist; *p != 0 && p <= last; p++) {
-			p[0] = p[1];
-		}
-	}
+	BaseMusic::SetSet(set_name.c_str());
+
+	this->BuildPlaylists();
+	this->ChangePlaylist(this->selected_playlist);
 
-	/* Make sure the list is null terminated. */
-	*last = 0;
+	InvalidateWindowData(WC_GAME_OPTIONS, WN_GAME_OPTIONS_GAME_OPTIONS, 0, true);
 }
 
-/** Prepare the playlists */
-void InitializeMusic()
+/** Enable shuffle mode and restart playback */
+void MusicSystem::Shuffle()
 {
-	uint j = 0;
-	for (uint i = 0; i < NUM_SONGS_AVAILABLE; i++) {
-		if (StrEmpty(GetSongName(i))) continue;
-		_playlist_all[j++] = i + 1;
-	}
-	/* Terminate the list */
-	_playlist_all[j] = 0;
-
-	/* Now make the 'styled' playlists */
-	for (uint k = 0; k < NUM_SONG_CLASSES; k++) {
-		j = 0;
-		for (uint i = 0; i < NUM_SONGS_CLASS; i++) {
-			int id = k * NUM_SONGS_CLASS + i + 1;
-			if (StrEmpty(GetSongName(id))) continue;
-			_playlists[k + 1][j++] = id + 1;
-		}
-		/* Terminate the list */
-		_playlists[k + 1][j] = 0;
+	_settings_client.music.shuffle = true;
+
+	this->active_playlist = this->displayed_playlist;
+	for (size_t i = 0; i < this->active_playlist.size(); i++) {
+		size_t shuffle_index = InteractiveRandom() % (this->active_playlist.size() - i);
+		auto &a = this->active_playlist[i];
+		auto &b = this->active_playlist[i + shuffle_index];
+		std::swap(a, b);
 	}
 
-	ValidatePlaylist(_settings_client.music.custom_1, lastof(_settings_client.music.custom_1));
-	ValidatePlaylist(_settings_client.music.custom_2, lastof(_settings_client.music.custom_2));
+	if (_settings_client.music.playing) this->Play();
 
-	if (BaseMusic::GetUsedSet()->num_available < _music_wnd_cursong) {
-		/* If there are less songs than the currently played song,
-		 * just pause and reset to no song. */
-		_music_wnd_cursong = 0;
-		_song_is_active = false;
-	}
+	InvalidateWindowData(WC_MUSIC_WINDOW, 0);
 }
 
-static void SkipToPrevSong()
+/** Disable shuffle and restart playback */
+void MusicSystem::Unshuffle()
 {
-	byte *b = _cur_playlist;
-	byte *p = b;
-	byte t;
+	_settings_client.music.shuffle = false;
+	this->active_playlist = this->displayed_playlist;
 
-	if (b[0] == 0) return; // empty playlist
+	if (_settings_client.music.playing) this->Play();
 
-	do p++; while (p[0] != 0); // find the end
+	InvalidateWindowData(WC_MUSIC_WINDOW, 0);
+}
 
-	t = *--p; // and copy the bytes
-	while (p != b) {
-		p--;
-		p[1] = p[0];
-	}
-	*b = t;
+/** Start/restart playback at current song */
+void MusicSystem::Play()
+{
+	/* Always set the playing flag, even if there is no music */
+	_settings_client.music.playing = true;
+	MusicDriver::GetInstance()->StopSong();
+	this->ChangePlaylistPosition(0);
```

Comment:
this is less obvious, shouldn't this do nothing?

Reply:
I'll add a comment. The intent is to make sure playlist_position is valid, in case playlist was changed to a shorter one etc.